File: leetcode/0128_longest_consecutive_sequence.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <map>
#include <utility>
#include <vector>

using namespace std;

class Solution {
 public:
    int longestConsecutive(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }

        map<int, int> lengths;
        for (auto num : nums) {
            lengths[num] = 0;
        }

        for (auto num : nums) {
            auto it = lengths.find(num);
            if (it->second == 0) {
                it->second = 1;

                int left;
                auto leftIt = lengths.find(num - 1);
                if (leftIt == lengths.end()) {
                    left = 0;
                } else {
                    left = leftIt->second;
                }

                int right;
                auto rightIt = lengths.find(num + 1);
                if (rightIt == lengths.end()) {
                    right = 0;
                } else {
                    right = rightIt->second;
                }

                int length = 1 + left + right;
                lengths[num - left] = length;
                lengths[num + right] = length;
            }
        }

        auto it = max_element(
            lengths.begin(), lengths.end(),
            [] (pair<int, int> p, pair<int, int> q) {
                return p.second < q.second;
            });
        return it->second;
    }
};

int main() {
    auto solution = Solution();

    vector<int> nums = {100, 4, 200, 1, 3, 2};

    assert(4 == solution.longestConsecutive(nums));
}

File: leetcode/0128_longest_consecutive_sequence.py
# -*- coding: utf-8 -*-


class Solution(object):
    def longestConsecutive(self, nums):
        if not nums:
            return 0

        lengths = {num: 0 for num in nums}
        for num in nums:
            if not lengths[num]:
                lengths[num] = 1
                left, right = lengths.get(num - 1, 0), lengths.get(num + 1, 0)
                length = 1 + left + right
                lengths[num - left], lengths[num + right] = length, length
        return max(lengths.values())


if __name__ == '__main__':
    solution = Solution()

    assert 4 == solution.longestConsecutive([100, 4, 200, 1, 3, 2])
    assert 0 == solution.longestConsecutive([])
    assert 3 == solution.longestConsecutive([1, 2, 0, 1])
